captum/optim/_utils/models.py

```diff
@@ -151,3 +155,72 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
             self.dilation,
             self.groups,
         )
+
+
+def pad_reflective_a4d(x: torch.Tensor, padding: List[int]) -> torch.Tensor:
+    """
+    Reflective padding for all 4 dimensions of an NCHW tensor
+    """
+
+    assert x.dim() == 4
+    assert len(padding) == 8
+
+    # Pad width
+    if padding[0] != 0:
+        x = torch.cat([x, x.flip([3])[..., 0 : padding[0]]], dim=3)
+    if padding[1] != 0:
+        x = torch.cat([x.flip([3])[..., -padding[1] :], x], dim=3)
+
+    # Pad height
+    if padding[2] != 0:
+        x = torch.cat([x, x.flip([2])[..., 0 : padding[2], :]], dim=2)
+    if padding[3] != 0:
+        x = torch.cat([x.flip([2])[..., -padding[3] :, :], x], dim=2)
+

```

Comment:
for the hight and width we can probably use built-in function:
torch.nn.functional.pad(x, padding, "reflect") 

As I was trying it out I noticed that some elements get repeated:
```
... [22., 21., 21.,  ..., 24., 24., 23.] ...
```
I think that it would make sense to test the code for a tensor generated with torch.arange and  asserting actual values.

Reply:
Looks like repeating the boundary elements is desirable as that's what TensorFlow's symmetric padding does: https://github.com/pytorch/pytorch/issues/46240

Reply:
@ProGamerGov, symmetric looks like replication padding but it is not implemented the way it is described in this article.
https://www.machinecurve.com/index.php/2020/02/07/what-is-padding-in-a-neural-network/

I think you can compare your implementation with the numpy variant to make sure that it works correctly
np.pad(a, 2, mode='symmetric')

Reply:
I think for the moment reflection padding should be okay. I modified the reflection padding function so that it works for all values, except an input size of 2 for channels and batch. https://github.com/ProGamerGov/captum/blob/optim-wip-patch-1/captum/optim/_utils/models.py#L160-L219

I'm currently trying to figure out how it's supposed to work with a size of 2.

Reply:
I compared my improved function with NumPy's reflection padding, and it works exactly the same (minus a batch or channel size of 2)!

Reply:
I see - Are we supporting symmetric padding as in lucid or the reflection padding ?https://github.com/tensorflow/lucid/blob/master/lucid/optvis/param/lowres.py#L63

Reply:
Lucent uses reflection padding in place of symmetric padding, so for the moment I think we can do reflection padding. In a future PR, we can implement symmetric padding as an option for SharedImage.

Reply:
I've also reached out to Chris to double check that reflection padding is okay to use.

Reply:
@ProGamerGov, let's keep the conversation in the PR thread in case other people look into this PR as well - I've received some of the messages as email response.

Reply:
@ProGamerGov wrote: 

>>> I've also reached out to Chris to double check that reflection padding is okay to use.

Sounds good, thank you!

Reply:
@NarineK I haven't sent any emails, so I assume you mean the comments I deleted? I deleted some comments because I felt like they either didn't make sense, were completely wrong, or other stuff of that nature.